src/metrics/mod.rs

```diff
@@ -118,3 +118,31 @@ pub fn capture_metrics(controller: &Controller) -> impl Iterator<Item = Event> {
 
     events.into_iter()
 }
+
+#[macro_export]
+/// This macro is used to emit metrics as a `counter` while simultaneously converting from absolute
+/// values to incremental values.
+///
+/// # Panics
+///
+/// If a subsequent value is smaller than the previous one.
+macro_rules! update_counter {
+    ($label:literal, $value:expr) => {{
+        use ::std::sync::Mutex;
+
+        ::lazy_static::lazy_static! {
+            static ref PREVIOUS_VALUE: Mutex<u64> = Mutex::new(0);
```

Comment:
Why a `Mutex` here and not an `AtomicU64`?

Reply:
If we only swap the previous value/new value, there would be no guarantee that the incremental values are emitted in the same order as the absolute ones:

Thread 1: Absolute value `5` comes in
Thread 1: Swaps previous value `0` with new value `5`
Thread 2: Absolute value `20` comes in
Thread 2: Swaps previous value `5` with new value `20`
Thread 2: Calculates the delta `15` and emits it
Thread 1: Calculates the delta `5` and emits it

Now I'm not sure if the order of the incremental values is critical behavior that needs to be preserved, any thoughts about that?

In any case, thanks for raising this question. This should be a deliberate choice if we go with the `Mutex`.

Reply:
I just noticed that I drop the lock guard before emitting the delta, so this definitely needs some change.

Reply:
I don't know that it matters for all practical purposes. No matter which order the deltas are created, the result on the receiver will have the same value once both are output, and they will be output in very short succession.

Reply:
Good questions @pablosichert and they make me see more why you started out with a mutex here. That makes me wonder if we might not do well to extend our `Recorder` to allow you to just set this kind of counter directly. The trouble that we've gotten ourselves into here is metrics-rs macros assume incremental updates but we have a source that just gives us the absolute, a gauge but not. May be we side-step this problem by letting a counter be updated like a gauge, with some added protections around breaking monotonic guarantees etc. Couldn't use the `counter!` macro anymore, unless this kind of change were of interest to upstream metrics-rs. /cc @tobz

Reply:
@blt Jep, I agree with you regarding the `metrics` crate and brought up the same points in https://github.com/metrics-rs/metrics/issues/172.

The essence of the argument why they don't provide such an interface was:

> If we allow absolute value updates, we can not provide that invariant.

For our use case the performance implications seem way worse than potentially breaking/ignoring an invariant, though. Augmenting our code base locally by an `update_counter` seems okay.

I'm just assuming that other users with the same requirements might hit that performance degradation when using the `AbsoluteLayer`. Which could be a good point to make when going back to that discussion, now that we have some actual numbers.

---

@bruceg since these metrics are set to emit only once a second by default, it doesn't seem to affect our practical purpose. Nevertheless an interesting discussion for when we chose to use that macro in a place with more throughput.

Reply:
Yeah, I'm really not willing to change the API here.  The `Absolute` layer covers this explicitly, although obviously it doesn't entirely make sense for applications that have very high metric throughput.  I'm open to thoughts on improving the performance, though.

As far as the logic mentioned above, the T1/T2 scenario is valid if you only use the semantics of `swap`, but you could trivially use a compare-and-swap loop to correctly replace the value and generate the correct delta.

Reply:
I just tried an approach to implement this using a compare exchange loop.

> but you could trivially use a compare-and-swap loop to correctly replace the value and generate the correct delta.

That is correct, but unless I am missing something we still can't make sure that _emitting_ the delta happens in order.

```rust
use ::std::sync::atomic::{AtomicU64, Ordering};

::lazy_static::lazy_static! {
    static ref PREVIOUS_VALUE: AtomicU64 = AtomicU64::new(0);
}

let mut previous_value = PREVIOUS_VALUE.load(Ordering::Relaxed);
let delta = loop {
    let delta = $value.checked_sub(previous_value);

    /// 1

    match PREVIOUS_VALUE.compare_exchange_weak(previous_value, $value, Ordering::SeqCst, Ordering::Relaxed) {
        Ok(_) => break delta,
        Err(value) => previous_value = value,
    }
};

/// 2
```

If we place

```rust
if delta.is_none() {
    warn!(message = "Violation of update_counter! invariant, called with a series of non-monotonic values.", label = $label);
}

counter!($label, delta.unwrap_or(0));
```

at `/// 1`, we might be too optimistic and emit it more than once, if we place it at `/// 2` we can't guarantee order.

(Also noticed that we don't need `compare_exchange` over `swap` here since we don't update the previous value with a calculated value.)

Reply:
I think what you want here is something like...

1.) read kafka value once (this is important)
2.) enter the loop
3.) read the current atomic value
4.) if the atomic value is greater than or equal to the kafka value, someone else beat you to the punch, all done
5.) otherwise, try to compare_exchange the atomic value with your kafka value
6.) if you win the exchange, take the delta of the previous/new values and emit, all done

Essentially, you end up letting a farther-in-the-future call win out over the an older call, and since only the caller that wins the exchange does the delta, you don't have to worry about incremental deltas or doing it in order.  You specifically avoid re-reading the source/kafka value, though, because you might end up doing extra work if the value is constantly changing in the exchange loop.  It's unlikely it changes fast enough to keep the loop going forever... but better to not risk it.

Reply:
My point is rather with `6.`: Winning the exchange and emitting does not happen atomically.

Thread 1: Reads kafka value `123`
Thread 2: Reads kafka value `456`
Thread 1: In loop `2.`
Thread 2: In loop `2.`
Thread 1: Reads atomic `0` in `3.`
Thread 1: Passes comparison `123 >= 0` in `4.`
Thread 1: Wins exchange in `6.` and writes `123`
Thread 2: Reads atomic `123` in `3.`
Thread 2: Passes comparison `456 >= 123` in `4.`
Thread 2: Wins exchange in `6.` and writes `456`
Thread 2: Emits delta `333`
Thread 1: Emits delta `123`

Where the correct order should have been `[123, 333]` or `456` if thread 2 arrived with its value before thread 1 and thread 1 failed the comparison at `4.`.

The question is if that order matters. The consensus here seems to be it does not?

Reply:
Ah, gotcha.  You're totally right, the emitting of the counter and the tracking of the delta isn't in lockstep, but practically speaking, it shouldn't need to be.  I agree with Bruce here.